test/test_hookcmds.py

```diff
@@ -0,0 +1,449 @@
+    assert result[0].port == 8080
+    assert result[0].protocol == 'tcp'
+
+
+def test_relation_get(run: Run):
+    run.handle(['relation-get', '--format=json'], stdout='{"foo": "bar"}')
+    result = hookcmds.relation_get()
+    assert result == {'foo': 'bar'}
+
+
+def test_relation_ids(run: Run):
+    run.handle(['relation-ids', 'db', '--format=json'], stdout='["rel:1"]')
+    result = hookcmds.relation_ids('db')
+    assert result == ['rel:1']
+
+
+def test_relation_list(run: Run):
+    run.handle(['relation-list', '--format=json'], stdout='["unit/0"]')
+    result = hookcmds.relation_list()
+    assert result == ['unit/0']
+
+
+def test_relation_model_get(run: Run):
+    data = {'uuid': str(uuid.uuid4())}
+    run.handle(['relation-model-get', '--format=json'], stdout=json.dumps(data))
+    result = hookcmds.relation_model_get()
+    assert str(result.uuid) == data['uuid']
+
+
+def test_relation_set(run: Run, mock_file: NamedTemporaryFile):
+    run.handle(['relation-set', '--file', '-'])
+    hookcmds.relation_set({'foo': 'bar'})
+    assert run.calls[0].stdin == 'foo: bar\n'
+
+
+def test_resource_get(run: Run):
+    run.handle(['resource-get', 'res'], stdout='/path/to/resource')
+    result = hookcmds.resource_get('res')
+    assert str(result) == '/path/to/resource'
+
+
+def test_secret_add(run: Run, mock_temp_dir: str):
+    run.handle(['secret-add', f'foo#file={mock_temp_dir}/foo'], stdout='secretid')
+    result = hookcmds.secret_add({'foo': 'bar'})
+    assert result == 'secretid'
+
+
+def test_secret_get(run: Run):
+    run.handle(['secret-get', '--format=json', 'secret:123'], stdout='{"foo": "bar"}')
+    result = hookcmds.secret_get(id='secret:123')
+    assert result == {'foo': 'bar'}
+
+
+def test_secret_grant(run: Run):
+    run.handle(['secret-grant', '--relation', '1', 'id'])
+    hookcmds.secret_grant('id', 1)
+
+
+def test_secret_ids(run: Run):
+    run.handle(['secret-ids', '--format=json'], stdout='["id1", "id2"]')
+    result = hookcmds.secret_ids()
+    assert result == ['id1', 'id2']
+
+
+def test_secret_info_get(run: Run):
+    info = {
+        '123': {
+            'label': 'lbl',
+            'description': 'desc',
+            'expiry': None,
+            'rotation': None,
+            'rotates': None,
+            'revision': 1,
+        }
+    }
+    run.handle(['secret-info-get', '--format=json', 'secret:123'], stdout=json.dumps(info))
+    result = hookcmds.secret_info_get(id='secret:123')
+    assert result.id == '123'
+    assert result.label == 'lbl'
+
+
+def test_secret_remove(run: Run):
+    run.handle(['secret-remove', 'id'])
+    hookcmds.secret_remove('id')
+
+
+def test_secret_revoke(run: Run):
+    run.handle(['secret-revoke', 'id'])
+    hookcmds.secret_revoke('id', relation_id=None, app=None)
+
+
+def test_secret_set(run: Run, mock_temp_dir: str):
+    run.handle(['secret-set', 'secret:123', f'foo#file={mock_temp_dir}/foo'])
+    hookcmds.secret_set('secret:123', content={'foo': 'bar'})
```

Comment:
Note from discussion with Dima, have both a minimal and an "all args" test for the more complex commands like this.